iree/samples/variables_and_state/main.c

```diff
@@ -0,0 +1,195 @@
+          driver_name.data);
+  iree_runtime_instance_options_t instance_options;
+  iree_runtime_instance_options_initialize(IREE_API_VERSION_LATEST,
+                                           &instance_options);
+  iree_runtime_instance_options_use_all_available_drivers(&instance_options);
+  iree_runtime_instance_t* instance = NULL;
+  IREE_RETURN_IF_ERROR(iree_runtime_instance_create(
+      &instance_options, iree_allocator_system(), &instance));
+  // TODO(#5724): move device selection into the compiled modules.
+  iree_hal_device_t* device = NULL;
+  IREE_RETURN_IF_ERROR(iree_runtime_instance_try_create_default_device(
+      instance, driver_name, &device));
+  //===-------------------------------------------------------------------===//
+
+  //===-------------------------------------------------------------------===//
+  // Session configuration (one per loaded module to hold module state).
+  fprintf(stdout, "Creating IREE runtime session\n");
+  iree_runtime_session_options_t session_options;
+  iree_runtime_session_options_initialize(&session_options);
+  iree_runtime_session_t* session = NULL;
+  IREE_RETURN_IF_ERROR(iree_runtime_session_create_with_device(
+      instance, &session_options, device,
+      iree_runtime_instance_host_allocator(instance), &session));
+  iree_hal_device_release(device);
+
+  fprintf(stdout, "Loading bytecode module at '%s'\n",
+          bytecode_module_path.data);
+  IREE_RETURN_IF_ERROR(iree_runtime_session_append_bytecode_module_from_file(
+      session, bytecode_module_path.data));
+  //===-------------------------------------------------------------------===//
+
+  //===-------------------------------------------------------------------===//
+  // Look up functions.
+  iree_runtime_call_t call_get_value;
+  iree_runtime_call_t call_set_value;
+  iree_runtime_call_t call_add_to_value;
+  iree_runtime_call_t call_reset_value;
+  // Exported function names match what we used in Python.
+  IREE_RETURN_IF_ERROR(iree_runtime_call_initialize_by_name(
+      session, iree_make_cstring_view("module.get_value"), &call_get_value));
+  IREE_RETURN_IF_ERROR(iree_runtime_call_initialize_by_name(
+      session, iree_make_cstring_view("module.set_value"), &call_set_value));
+  IREE_RETURN_IF_ERROR(iree_runtime_call_initialize_by_name(
+      session, iree_make_cstring_view("module.add_to_value"),
+      &call_add_to_value));
+  IREE_RETURN_IF_ERROR(iree_runtime_call_initialize_by_name(
+      session, iree_make_cstring_view("module.reset_value"),
+      &call_reset_value));
+  //===-------------------------------------------------------------------===//
+
+  //===-------------------------------------------------------------------===//
+  // Call functions to manipulate the counter
+  fprintf(stdout, "Calling functions\n\n");
+
+  // 1. get_value() // initial value
+  fprintf(stdout, "Initial get_value():\n");
+  IREE_RETURN_IF_ERROR(counter_print_current_value(&call_get_value));
```

Comment:
nit: I think to match the python more you should just inline the print - the counter type has a get_value, and then you pass the result of that to print() - this kind of loses the connection to the original model by wrapping it

Reply:
Agreed. Changed the call signature to 
```c
iree_status_t counter_get_value(iree_runtime_session_t* session,
                                int* out_value) {
```
and used `iree_hal_buffer_map_range` to get the value (instead of using `iree_hal_buffer_view_fprint`) then inlined the prints.

Output now looks like this:
```
λ ..\iree-build\iree\samples\variables_and_state\variables-and-state.exe D:\dev\projects\iree-tmp\counter_vmvx.vmfb vmvx
Configuring IREE runtime instance and 'vmvx' device
Creating IREE runtime session
Loading bytecode module at 'D:\dev\projects\iree-tmp\counter_vmvx.vmfb'
Calling functions

Initial get_value()    : 0
After set_value(101)   : 101
After add_to_value(20) : 121
After add_to_value(-50): 71
After reset_value()    : 0

Success!
```